sbnci/Modules/TrackValidation/TrackValidation_module.cc

```diff
@@ -0,0 +1,3127 @@
+  //###### Declaration and Initiate variables #######
+  for(auto const& mapiter : TrueParticlesMap){
+    //Declaration and Initiate variables
+    //
+    std::vector<unsigned int> ParTrkPdgCode = fTrackPdgCodeVect;//{13,211,2212,321} PdgCode of track like particles
+    //
+    const simb::MCParticle *particle = mapiter.second;
+    // looking at the PdgCode of the particles
+    unsigned int particlePdg = std::abs(particle->PdgCode());
+    // Make sure that we are looking at track particles and daughter
+    if(std::find(ParTrkPdgCode.begin(), ParTrkPdgCode.end(), particlePdg) == ParTrkPdgCode.end()){continue;}
+    //Check the mother is not a track particle
+    if(TrueParticlesMap.find(particle->Mother()) != TrueParticlesMap.end()){
+      if(std::find(ParTrkPdgCode.begin(), ParTrkPdgCode.end(), (std::abs(TrueParticlesMap[particle->Mother()]->PdgCode()))) != ParTrkPdgCode.end()){
+        //order the daughter based on it's mother.
+        int MotherId = particle->Mother();
+        int part_temp = MotherId;
+        //Match particle with mother
+        while(MotherId != 0){
+          part_temp = TrueParticlesMap[MotherId]->Mother();
+          if(TrueParticlesMap.find(part_temp) == TrueParticlesMap.end()){break;}
+          if(std::find(ParTrkPdgCode.begin(), ParTrkPdgCode.end(), (std::abs(TrueParticlesMap[part_temp]->PdgCode()))) == ParTrkPdgCode.end()){break;}
+          MotherId = part_temp;
+        }
+        //Add to the mother chain.
+        MothTrkIdWithDauIdsMap[MotherId].push_back(particle->TrackId());
+        continue;
+      }
+    }//close the first condition block
+    //
+    MothTrkIdWithDauIdsMap[particle->TrackId()].push_back(particle->TrackId());
+  }//close the loop over true mc particles
+  return;
+}//close MotherTrackIdsTodauIdsFun()
+
+// Function retrun the true track id
+int ana::TrackValidation::FunTrueParticleID(detinfo::DetectorClocksData const& clockData,
+                                            const art::Ptr<recob::Hit>& hit) const{
+  //###### Declaration and Initiate variables #######
+  std::map<int,double> id_to_energy_map;
+  double particleEnergy = -99999;
+  int likelyTrackID = 0;
+  std::vector<sim::TrackIDE> trackIDs = bt_serv->HitToTrackIDEs(clockData, hit);
+  for(unsigned int idIt = 0; idIt < trackIDs.size(); ++idIt){
+    //if(trackIDs.at(idIt).energy > particleEnergy){
+    double Energy = trackIDs.at(idIt).energy;  
+    int TrkID = std::abs(trackIDs.at(idIt).trackID);
+    id_to_energy_map[TrkID] += Energy;
+  }
+  //
+  for(auto const& mapIter : id_to_energy_map){
+    double energy_map = mapIter.second;
+    if(energy_map > particleEnergy){
+      particleEnergy = energy_map;
+      likelyTrackID = mapIter.first;
+    }
+  }//close the loop over the map
+  return likelyTrackID;
+}//Close the function retrun the true track id
+
+//Function return all true hits of a particle
+void ana::TrackValidation::TruthTrkIdsToHits(art::Event const& evt,
```

Comment:
Is there a reason not to use TruthMatchUtils: https://github.com/LArSoft/larsim/blob/develop/larsim/Utils/TruthMatchUtils.h

Reply:
resolved